web/client/src/library/components/table/Table.tsx

```diff
@@ -0,0 +1,256 @@
+  ChevronUpIcon,
+} from '@heroicons/react/24/solid'
+import { useVirtual } from '@tanstack/react-virtual'
+import Input from '@components/input/Input'
+import { EnumSize } from '~/types/enum'
+import { isArrayNotEmpty } from '@utils/index'
+
+declare module '@tanstack/table-core' {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  interface ColumnMeta<TData extends RowData, TValue> {
+    type: string
+  }
+}
+
+const MIN_HEIGHT_ROW = 24
+
+export default function Table({
+  data = [[], []],
+}: {
+  data: [TableColumn[], TableRow[]]
+}): JSX.Element {
+  const elTableContainer = useRef<HTMLDivElement>(null)
+
+  const columns = useMemo(
+    () =>
+      data[0].map(({ name, type }) => ({
+        accessorKey: name,
+        meta: {
+          type,
+        },
+      })),
+    [data[0]],
+  )
+
+  const [sorting, setSorting] = useState<SortingState>([])
+  const [filter, setFilter] = useState('')
+
+  const table = useReactTable({
+    data: data[1],
+    columns,
+    state: {
+      sorting,
+      globalFilter: filter,
+    },
+    onGlobalFilterChange: setFilter,
+    onSortingChange: setSorting,
+    getCoreRowModel: getCoreRowModel(),
+    getSortedRowModel: getSortedRowModel(),
+    getFilteredRowModel: getFilteredRowModel(),
+  })
+
+  const { rows } = table.getRowModel()
+  const rowVirtualizer = useVirtual({
+    parentRef: elTableContainer,
+    size: rows.length,
+    overscan: 10,
+  })
+  const { virtualItems: virtualRows, totalSize } = rowVirtualizer
+
+  const paddingTop = virtualRows.length > 0 ? virtualRows?.[0]?.start ?? 0 : 0
+  const paddingBottom =
+    virtualRows.length > 0
+      ? totalSize - (virtualRows?.[virtualRows.length - 1]?.end ?? 0)
+      : 0
+
+  return (
+    <div className="w-full h-full flex flex-col">
+      <Header
+        filter={filter}
+        setFilter={setFilter}
+      />
+      <div
+        ref={elTableContainer}
+        className="w-full h-full rounded-lg overflow-auto hover:scrollbar scrollbar--horizontal scrollbar--vertical"
+      >
+        <table
+          cellPadding={0}
+          cellSpacing={0}
+          className={clsx(
+            'w-full slashed-zero tabular-nums',
+            'text-neutral-700 dark:text-neutral-300 text-xs font-medium whitespace-nowrap text-left',
+          )}
+        >
+          <thead className="sticky top-0">
+            {table.getHeaderGroups().map(headerGroup => (
+              <tr
+                key={headerGroup.id}
+                className="bg-primary-10 dark:bg-secondary-10 backdrop-blur-lg"
+                style={{ height: `${MIN_HEIGHT_ROW}px` }}
+              >
+                {headerGroup.headers.map(header => (
+                  <th
+                    key={header.id}
+                    className={clsx(
+                      'pl-2 pr-4 pt-2 text-sm pb-1 border-r-2 last:border-r-0 border-light dark:border-dark',
+                    )}
+                  >
+                    {header.isPlaceholder ? null : (
```

Comment:
If it's a placeholder, does the word `null` appear on the screen?

Reply:
no